### PR TITLE
Uniswap route

### PR DESCRIPTION
**Pull request by piyush2-0**
Added uniswap route on:-
- [x] Optimism
- [x] Polygon
- [x] Arbitrum

### CODE REVIEW

File: contracts/aggregator/optimism/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+     * @param _tokens list of token addresses.
+     */
+    function validateTokens(address[] memory _tokens) internal pure {
+        for (uint256 i = 0; i < _tokens.length - 1; i++) {
+            require(_tokens[i] != _tokens[i + 1], "non-unique-tokens");
+        }
+    }
+
+
+    function setPoolAddress(address pool) internal {
```
Comment (Vaibhav20201) [inline contracts/aggregator/optimism/helpers.sol:161]: Remove this

File: contracts/aggregator/optimism/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        uniswapPoolAddress = pool;
+    }
+
+    /**
+     * @dev Returns fee for the passed route in BPS.
+     * @notice Returns fee for the passed route in BPS. 1 BPS == 0.01%.
+     * @param _route route number for flashloan.
+     */
+    function calculateFeeBPS(uint256 _route)
```
Comment (Vaibhav20201) [inline contracts/aggregator/optimism/helpers.sol:170]: Can remove for optimism

File: contracts/aggregator/optimism/main.sol [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        instaLoanVariables_._amounts = amounts_;
+        instaLoanVariables_._iniBals = calculateBalances(
+            tokens_,
+            address(this)
+        );
+
+        setPoolAddress(uniswapPoolAddress);
+        instaLoanVariables_._instaFees = calculateFees(
+            amounts_,
+            calculateFeeBPS(8)
```
Comment (Vaibhav20201) [inline contracts/aggregator/optimism/main.sol:90]: Directly pass from pool key after factoring decimals

**Review by Vaibhav20201 — CHANGES_REQUESTED** [review]

File: contracts/aggregator/optimism/main.sol [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        } else if (_route == 3) {
+            revert("this route is only for mainnet");
+        } else if (_route == 4) {
+            revert("this route is only for mainnet");
+        } else if (_route == 5) {
+            revert("this route is  not for optimism");
+        } else if (_route == 6) {
+            revert("this route is only for mainnet");
+        } else if (_route == 7) {
+            revert("this route is  not for optimism");
```
Comment (Vaibhav20201) [inline contracts/aggregator/optimism/main.sol:208]: Lets remove this part.

File: contracts/aggregator/optimism/flashloan/main.sol [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+     * @param _amounts list of amounts for the corresponding assets.
+     * @param _route route for flashloan.
+     * @param _data extra data passed.
+     */
+    function flashLoan(
+        address[] memory _tokens,
+        uint256[] memory _amounts,
+        uint256 _route,
+        bytes calldata _data,
+        bytes calldata _instadata // kept for future use by instadapp. Currently not used anywhere.
```
Comment (KaymasJain) [inline contracts/aggregator/optimism/flashloan/main.sol:181]:
```suggestion
        bytes calldata _instadata
```

**Review by KaymasJain — APPROVED** [review]

**Review by Vaibhav20201 — APPROVED** [review]

**Pull request merged**